python/ray/data/_internal/execution/interfaces/op_runtime_metrics.py

```diff
@@ -12,266 +12,289 @@
     )
 
 
+# A metadata key used to mark a dataclass field as a metric.
+_IS_FIELD_METRIC_KEY = "__is_metric"
+# Metadata keys used to store information about a metric.
+_METRIC_FIELD_DESCRIPTION_KEY = "__metric_description"
+_METRIC_FIELD_METRICS_GROUP_KEY = "__metric_metrics_group"
+_METRIC_FIELD_IS_MAP_ONLY_KEY = "__metric_is_map_only"
+
+_METRICS: List["Metric"] = []
+
+
+@dataclass(frozen=True)
+class Metric:
+    """Metadata for a metric.
+
+    Args:
+        name: The name of the metric.
+        description: A human-readable description of the metric, also used as the chart
+            description on the Ray Data dashboard.
+        metrics_group: The group of the metric, used to organize metrics into groups in
+            'StatsActor' and on the Ray Data dashboard.
+        map_only: Whether the metric is only measured for 'MapOperators'.
+    """
+
+    name: str
+    description: str
+    metrics_group: str
+    map_only: bool = False
+
+
+def metricfield(
+    *,
+    description: str,
+    metrics_group: str,
+    map_only: bool = False,
+    **field_kwargs,
+):
+    """A dataclass field that represents a metric."""
+    metadata = field_kwargs.get("metadata", {})
+
+    metadata[_IS_FIELD_METRIC_KEY] = True
+
+    metadata[_METRIC_FIELD_DESCRIPTION_KEY] = description
+    metadata[_METRIC_FIELD_METRICS_GROUP_KEY] = metrics_group
+    metadata[_METRIC_FIELD_IS_MAP_ONLY_KEY] = map_only
+
+    return field(metadata=metadata, **field_kwargs)
+
+
+def metricproperty(
+    *,
+    description: str,
+    metrics_group: str,
+    map_only: bool = False,
+):
+    """A property that represents a metric."""
+
+    def wrap(func):
+        metric = Metric(
+            name=func.__name__,
+            description=description,
+            metrics_group=metrics_group,
+            map_only=map_only,
+        )
+
+        _METRICS.append(metric)
+
+        return property(func)
+
+    return wrap
+
+
 @dataclass
 class RunningTaskInfo:
     inputs: RefBundle
     num_outputs: int
     bytes_outputs: int
 
 
+class OpRuntimesMetricsMeta(type):
```

Comment:
is it possible to just add the metric to the list in `metricfield`. So we don't need this meta class?

Reply:
alternatively, i think we can add that logic in `OpRuntimeMetrics.__post_init__`

Reply:
How would we get the name of the attribute in `metricfield`? `Field.name` is `None` in `metricfield`. I'm not sure how `name` exactly gets set, but I think it's some dataclass magic that occurs after the class is defined.

One alternative is to explicitly specify the name. Advantage is that we don't have to use metaclasses; disadvantage is that you need to duplicate the name:
```python
num_inputs_received: int = metricfield(
        name="num_inputs_received",
        default=0,
        description="Number of input blocks received by operator.",
        metrics_group="inputs",
    )
```

Reply:
personally, i prefer to only specify the name only when needed. if you put the logic in `__post_init__`, that should have a valid `name` attribute i think?

Reply:
> alternatively, i think we can add that logic in OpRuntimeMetrics.__post_init__

We could do this. We'd just need to make sure we don't add duplicate items since __post_init__ is called once per instance and not once when the class is defined.

Reply:
Don't have a strong preference for `post_init` vs. metaclass. Down for either

Reply:
Discussed offline, we propose putting the logic in `post_init`, and making `_METRICS` a set so that we don't keep duplicates.

Reply:
After testing it out, I think `__post_init__` might not work. `__post_init__` only works if you don't override `__init__`, and we currently override `__init__` to pass the `Operator` to the `OpRuntimeMetrics` instance.